### PR TITLE
Fix bug in TopK aggregates

### LINKED ISSUE #12748: Probable bug in TopKAggregate
### Describe the bug

@dispanser pointed out a probable bug in TopKAggregate:

https://github.com/apache/datafusion/blob/77f330c6a2b26f2d1d4d4bf11d456fad466316b4/datafusion/physical-optimizer/src/topk_aggregation.rs#L102

1. we should probably call it `cardinality_reducing` and allow anything that is `!cardinality_reducing`
2. `FilterExec` is cardinality reducing, and should not be whitelisted here
3. things like `CoalescePartitionsExec` should be white listed because they don't change cardinality
4. in theory, something like a left outer join can only increase cardinality, so should be allowed as well

I would suggest we add a `cardinality_reducing()` or `not_cardinality_reducing()` method to the `ExecutionPlan` trait, so we don't need to maintain this downcast list.

i.e. 
```
                    GlobalLimitExec: skip=0, fetch=100, metrics=[]
                      SortExec: TopK(fetch=100), expr=[$d.max_ts@1 DESC NULLS LAST], preserve_partitioning=[false], metrics=[]
                        AggregateExec: mode=Final, gby=[$d.traceID@0 as $d.traceID], aggr=[$d.max_ts], lim=[100], metrics=[]
                          CoalescePartitionsExec, metrics=[]
                            AggregateExec: mode=Partial, gby=[$d.traceID@0 as $d.traceID], aggr=[$d.max_ts], lim=[100], metrics=[]
```

is fine because `CoalescePartitionsExec` means the `100` rows from `AggregateExec` will be passed all the way up to `GlobalLimitExec`.

However:

```
                    GlobalLimitExec: skip=0, fetch=100, metrics=[]
                      SortExec: TopK(fetch=100), expr=[$d.max_ts@1 DESC NULLS LAST], preserve_partitioning=[false], metrics=[]
                        AggregateExec: mode=Final, gby=[$d.traceID@0 as $d.traceID], aggr=[$d.max_ts], lim=[100], metrics=[]
                          FilterExec, expr=[id % 2 == 0]
                            AggregateExec: mode=Partial, gby=[$d.traceID@0 as $d.traceID], aggr=[$d.max_ts], lim=[100], metrics=[]
```

Is not because `AggregateExec` could return `100` rows, `FilterExec` could make that `50`, then `GlobalLimitExec` doesn't get the `100` it was guaranteed in the original plan.

### To Reproduce

We need a new test.

### Expected behavior

Don't lose rows.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by avantgardnerio**
## Which issue does this PR close?

Closes #12748.

## Rationale for this change

We don't want bugs.

## What changes are included in this PR?

1. Push down agg limits past projections so we can observe the bug
2. Fix the bug

## Are these changes tested?

Yes

## Are there any user-facing changes?

1. queries that weren't optimized should now be optimized
2. queries that are optimized should return correct results

### CODE REVIEW

File: datafusion/physical-plan/src/sorts/sort.rs [hunk 1]
```diff
@@ -972,6 +973,10 @@ impl ExecutionPlan for SortExec {
     fn fetch(&self) -> Option<usize> {
         self.fetch
     }
+
+    fn cardinality_effect(&self) -> CardinalityEffect {
```
Comment (thinkharderdev) [inline datafusion/physical-plan/src/sorts/sort.rs:977]: Wouldn't this only be `Equal` if there is no limit? E.g. should it be 
```
if self.fetch.is_none() {
  CardinalityEffect::Equal
} else {
  CardinalityEffect::LowerEqual
}
```

File: datafusion/physical-plan/src/execution_plan.rs [hunk 2]
```diff
@@ -898,6 +903,13 @@ pub fn get_plan_string(plan: &Arc<dyn ExecutionPlan>) -> Vec<String> {
     actual.iter().map(|elem| elem.to_string()).collect()
 }
 
+pub enum CardinalityEffect {
+    Unknown,
+    Equal,
+    LowerEqual,
+    GreaterEqual,
+}
```
Comment (thinkharderdev) [inline datafusion/physical-plan/src/execution_plan.rs:911]:
```suggestion
/// Indicates the effect an execution plan operator will have on the cardinality
/// of its input stream
pub enum CardinalityEffect {
    /// Unknown effect. This is the default
    Unknown,
    /// The operator is guaranteed to produce exactly one row for 
    /// each input row
    Equal,
    /// The operator may produce fewer output rows than it receives input rows
    LowerEqual,
    /// The operator may produce more output rows than it receives input rows
    GreaterEqual,
}
```

**Review by Dandandan — APPROVED** [review]
Looks good to me 🚀

**Review by thinkharderdev — APPROVED** [review]
lgtm!

**Pull request merged**